Add support for css content-visibility: hidden

cathiechen opened the pull request:
#### faf9ff8c4023089d5aa557da0beb8b2737a2bc8b
<pre>
Add support for css content-visibility: hidden
<a href="https://bugs.webkit.org/show_bug.cgi?id=236710">https://bugs.webkit.org/show_bug.cgi?id=236710</a>

Reviewed by Antti Koivisto.

This patch implements support for content-visibility: hidden [1].
It introduces the fake inherited property effectiveSkipsContent to keep
track for Elements whether content-visibility is hidden and whether elements
should be skipped for painting and hit testing.

Several points in 4.3. Restrictions and Clarifications [2] are addressed too:
- scrollIntoView is adapted [4.3.5]
- tab-order navigation and focusing take skipped contents into account [3].

[1] <a href="https://drafts.csswg.org/css-contain/#using-cv-hidden">https://drafts.csswg.org/css-contain/#using-cv-hidden</a>
[2] <a href="https://drafts.csswg.org/css-contain/#cv-notes">https://drafts.csswg.org/css-contain/#cv-notes</a>
[3] <a href="https://drafts.csswg.org/css-contain/#valdef-content-visibility-hidden">https://drafts.csswg.org/css-contain/#valdef-content-visibility-hidden</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-015-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-016-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-017-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-018-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-029-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-035-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-038-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-047-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-072-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-img-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-input-image-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-svg-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-embed-element/embed-document-under-content-visibility-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-embed-element/embed-document-under-content-visibility-gbcr-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isFocusable const): skipped contents are not focusable
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollRectToVisible): do not scroll if the RenderObject is skipped content
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paintContents): do not paint contents for c-v: hidden
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::visibleToHitTesting const): mark skipped content as not visible to hit testing
(WebCore::RenderElement::shouldApplyLayoutContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyPaintContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyLayoutOrPaintContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplySizeContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyInlineSizeContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplySizeOrInlineSizeContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyStyleContainment const): c-v: hidden forces containment
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintList): do not paint child layers of a c-v: hidden container
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAfterDescendants):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContent const):
(WebCore::RenderObject::shouldSkipContent const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint): do not paint contents for content-visibility: hidden
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::paint): do not paint contents for content-visibility: hidden
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::effectiveSkipsContent const):
(WebCore::RenderStyle::setEffectiveSkipsContent):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/faf9ff8c4023089d5aa557da0beb8b2737a2bc8b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95029 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4172 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27939 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104631 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164884 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99019 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4261 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32356 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87335 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100532 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100696 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3094 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81548 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30074 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85007 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84654 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72970 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38762 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36585 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/4273 "The change is no longer eligible for processing.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40518 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42364 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42497 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38910 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/dom/Element.h:
```diff
@@ -100,6 +100,12 @@ struct ShadowRootInit;
+    WTF_MAKE_STRUCT_FAST_ALLOCATED;
+    bool isSkippedContent { false };
+    ContentVisibility contentVisibility;
+};
```
If there is nothing else to put here, this doesn't really need a heap allocated struct. Those fields could just sit directly in the rare data and take less space.

cathiechen replied:
Hmm, we will have more flags here for `content-visibility: auto` to indicate if the element is [relevant to the user](https://w3c.github.io/csswg-drafts/css-contain-2/#relevant-to-the-user), something like:
```
    bool isIntersectingViewport { false };
    bool subtreeHasFocus { false };
    bool subtreeHasSelection { false };
```
See the WIP-patch in https://bugs.webkit.org/show_bug.cgi?id=236711, looks like there isn't complex function there.
I'm fine to put `isSkippedContent` and `contentVisibility` in `ElementRareData`.

rwlbuis replied:
Hi Cathie, I think the idea was to get rid of the ContentVisibilityData completely, or at least not dynamically allocate it. Anyway I can fix this tomorrow morning.

rwlbuis replied:
I pushed a new version for what I think was the initial suggestion.

rwlbuis replied:
The failing win test has been flaky for a while so I think we can ignore it:
https://ews-build.webkit.org/#/builders/10

So this is ready for review again.

anttijk commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3389,7 +3389,7 @@ void RenderLayer::paintList(LayerList layerIterator, GraphicsContext& context, c
         return;
 
-    if (!hasSelfPaintingLayerDescendant())
+    if (!hasSelfPaintingLayerDescendant() || renderer().shouldSkipContent())
```
How is it determined where these tests are needed? It seems bit random.

The spec says "...to a large extent do not update their styles at all unless explicitly requested by script" and "user agent should additionally avoid as much layout/rendering work as possible for skipped contents" yet I see only tests about skipping painting and hit testing. This won't really help performance much. Is the plan to add those things later?

The spec also says "This is similar to giving the contents display: none". In display:none case we don't actually build a render tree at all for that subtree. Is that the intention of the spec?

cathiechen replied:
OK, I'll let @rwlbuis explain this, he does most work of this feature:)

rwlbuis replied:
The tests are added in order to comply with the specification/tests or to make things more efficient, see also the per method comments in the commit message.

A lot of the performance comes from enabling all kinds of containment when content-visibility is hidden, we did that by adjusting shouldApply*Containment methods. But note paint containment mostly works through overflow: clip, so this patch optimises more by avoiding painting of contents/children. I agree for hit testing it may not be a big win but for correctness we have to implement that (whenever the contents are skipped).

Regarding display: none, there are a lot of similarities, and I would not be surprised that its problems was the starting point for content-visibility. Basically display: none changes are heavy while content-visibility changes should be instant ideally. I think the hint was a way to help the reader think about content-visibility: hidden. In the content-visibility case I think we do want to build the render tree, to be less costly when updating the property, and to make content-visibility: auto efficient as well.

Work on top, besides the obvious content-visibility: auto work, will be in the area of editing, resize/intersection observer, top layers/dialogs, accessibility. From the top of my head, except maybe resize/intersection observer, I think it is more about correctness/spec compliance than performance.

anttijk commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -327,6 +327,8 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
         element.setDisplayContentsChanged(false);
     }
 
+    element.updateContentVisibility(elementUpdateStyle.contentVisibility());
```
RenderTreeUpdater putting style data into DOM like this is pretty nasty. I think this feature is best implemented as a fake inherited property in RenderStyle, similar to effectiveInert.

anttijk commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -521,37 +524,37 @@ inline bool RenderElement::shouldApplySizeOrStyleContainment(bool containsAccord
 inline bool RenderElement::shouldApplyLayoutContainment() const
 {
-    return shouldApplyLayoutOrPaintContainment(style().containsLayout());
+    return shouldApplyLayoutOrPaintContainment(style().containsLayout() || style().contentVisibility() != ContentVisibility::Visible);
```
Maybe this test could be in containsLayout() function? Same for containsPaint() etc.

rwlbuis replied:
I prefer to keep this as-is for now. Future work on this (forced layout/content-visibility: auto) may need to change/extend these code sections and it is not quite clear yet how.

anttijk commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -2553,6 +2553,16 @@ String RenderObject::debugDescription() const
 
+bool RenderObject::isSkippedContent() const
+{
+    return style().isInSkippedContentSubtree() && parent() && parent()->style().isInSkippedContentSubtree();
```
I think it is sufficient to test the parent bit only (render is skipped if the parent is skipping the content)

rwlbuis replied:
Nice thanks!

anttijk commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -550,6 +550,8 @@ class RenderStyle {
 
     ContentVisibility contentVisibility() const { return static_cast<ContentVisibility>(m_rareNonInheritedData->contentVisibility); }
 
+    bool isInSkippedContentSubtree() const { return m_rareInheritedData->isInSkippedContentSubtree; }
```
Something like effectiveSkipsContent might be less clunky and communicate that this is an effective property (similar to others).

rwlbuis replied:
Done.

anttijk reviewed (approved):
much better!

rwlbuis committed:
Add support for css content-visibility: hidden
https://bugs.webkit.org/show_bug.cgi?id=236710

Reviewed by Antti Koivisto.

This patch implements support for content-visibility: hidden [1].
It introduces the fake inherited property effectiveSkipsContent to keep
track for Elements whether content-visibility is hidden and whether elements
should be skipped for painting and hit testing.

Several points in 4.3. Restrictions and Clarifications [2] are addressed too:
- scrollIntoView is adapted [4.3.5]
- tab-order navigation and focusing take skipped contents into account [3].

[1] https://drafts.csswg.org/css-contain/#using-cv-hidden
[2] https://drafts.csswg.org/css-contain/#cv-notes
[3] https://drafts.csswg.org/css-contain/#valdef-content-visibility-hidden

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-015-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-016-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-017-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-018-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-029-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-035-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-038-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-047-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-072-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-img-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-input-image-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-svg-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-006-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-010-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-embed-element/embed-document-under-content-visibility-focus-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/embedded-content/the-embed-element/embed-document-under-content-visibility-gbcr-expected.txt:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::isFocusable const): skipped contents are not focusable
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollRectToVisible): do not scroll if the RenderObject is skipped content
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paintContents): do not paint contents for c-v: hidden
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::visibleToHitTesting const): mark skipped content as not visible to hit testing
(WebCore::RenderElement::shouldApplyLayoutContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyPaintContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyLayoutOrPaintContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplySizeContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyInlineSizeContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplySizeOrInlineSizeContainment const): c-v: hidden forces containment
(WebCore::RenderElement::shouldApplyStyleContainment const): c-v: hidden forces containment
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintList): do not paint child layers of a c-v: hidden container
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAfterDescendants):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContent const):
(WebCore::RenderObject::shouldSkipContent const):
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::paint): do not paint contents for content-visibility: hidden
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::paint): do not paint contents for content-visibility: hidden
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::effectiveSkipsContent const):
(WebCore::RenderStyle::setEffectiveSkipsContent):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
